### ISSUE TITLE
firstorder handles `prod A B` and `@sigT A (fun _ => B)` differently

### ISSUE THREAD

**Issue by jashug** [coq/coq#6393]
#### Version
The Coq Proof Assistant, version 8.8+alpha (December 2017)
compiled on Dec 12 2017 11:03:09 with OCaml 4.02.3

#### Operating system
Windows Subsystem for Linux (WSL) == Ubuntu 16.04

#### Description of the problem

The firstorder tactic gives different results when `prod A B` is replaced by `@sigT A (fun _ => B)`.
Example,
```
Goal prod True True. firstorder. Qed.
Goal @sigT True (fun _ => True). firstorder. Fail Qed. Abort.
Goal True -> @sigT True (fun _ => True). firstorder. Qed.
```
The first goal succeeds, while the second fails, leaving a goal of `True`, which firstorder should be able to solve. The third goal succeeds.
Some experimentation suggests that firstorder will only solve the first argument of sigT if
there is a hypothesis of the same type in the context.

This issue complicates unifying prop and sigT as proposed in #4110.

**Comment by jashug** [comment]
`tauto` and `dtauto` also behave differently for `prod` and `sigT`, solving `prod True True, and failing to solve either `@sigT True (fun _ => True)` or `True -> @sigT True (fun _ => True)`.

I tracked the divergence down to the behavior of the functions in [tactics/hipattern.ml](https://github.com/coq/coq/blob/master/tactics/hipattern.ml). The `has_nodep_prod_after` function there uses `Vars.noccurn`, which says `(fun _ : True => True) a` is dependent on `a` while `True` is not.

For conjunction, tauto/dtauto enters hipattern through match_with_conjunction, while firstorder enters through match_with_nodep_ind. Both call has_nodep_prod_after.

It seems that this behavior could be changed by normalizing the type of the constructor before checking for dependencies.

**Comment by herbelin** [comment]
Well spotted. Adding a `whd_all` before the `noccurn` could be tempted. The only risk I see is introducing costly computations, but it does not seem common to build very large types by reduction. Also, `whd_all` (= weak-head reduction) would be a good compromise in comparison to a full `nf_all` (= strong reduction to normal form) whose complexity can be arbitrary large (*).

As far as I'm concerned, I would be positive if you'd like to propose a PR improving on that.

(*) Ideally one would probably want to have a `nf_all` by default together with a tactic modifier, say, `freezing const-name in tac`, or `opaque const-name-list in tac` to handle the rare (though existing) cases where a `nf_all` would be too costly. But this goes beyond the scope of this issue strictly speaking.

**Comment by SkySkimmer** [comment]
NB: For this situation we don't need to delta reduce.

**Comment by herbelin** [comment]
> NB: For this situation we don't need to delta reduce.

Right. I was thinking at the possibility of say, a user using some `lift A` with `lift A := fun _ => A` but this is probably uncommon (or even, if so, it might be on purpose to create a phantom type). So, yes `whd_beta` could be the proper risk-free expected reduction to use.

**Comment by jashug** [comment]
I am currently working on a PR for this. The way `has_nodep_prod_after` currently works, is given the type of the pairing constructor:
`forall (a : True), forall (b : (fun  _ => True) a), @sigT True (fun _ => True)`,
it (recursively) breaks it into the variable a and the type `forall (b : (fun _ => True) a), ...`,
and checks that a does not occur in the latter type. Note that the latter type is already in weak head normal form (as I understand it); to reduce `(fun _ => True) a` we need to use at least `nf_beta`.

Furthermore, since we would be repeatedly normalizing the tail of the type, it seems to make more sense to normalize the type once using `nf_beta` (or something stronger), and then running `has_nodep_prod_after`.

Another approach would be to change the definition of `has_nodep_prod_after` to check that none of the previous variables appear in each (`whd_beta` reduced) type in the string of foralls. This would reduce the time we spend walking over the type from O(n^2) to O(n), and allow us to use `whd_beta` instead of `nf_beta`, but would be a much more invasive change to the code (smaller change in semantics).

To summarize, I see two options:
1. Insert a call to `nf_beta` at the three call sites of `has_no_dep_prod/has_no_dep_prod_after`,
    in `match_with_one_constructor`, `match_with_nodep_ind`, and `match_with_sigma_type`.
    This should fix the immediate problem, at a cost of calling `nf_beta` a few times when
    `tauto` or `firstorder` are called.
2. Change `has_nodep_prod_after` to eliminate the overlapping calls to `Vars.noccurn`,
     and insert `whd_beta` into the new version. This should have a much smaller impact on performance,
     (possibly even positive), but since this is my first time exploring the Coq source, I would need
    more guidance to make sure I preserve the semantics of `has_nodep_prod_after`.

I'll experiment with both, and submit PRs if I get things working.

**Comment by jashug** [comment]
This is unfortunately a bit more complicated than I thought.
* Adding normalization lets `dtauto` work with `@sigT A (fun _ => B)`. Yay!
* `firstorder` uses the `match_with_nodep_ind` path, which calls `has_no_dep_prod_after` on
  `forall A P (a : A) (b : P a), @sigT A P`, even though it has the actual values of `A` and `P` available.
  This seems changeable, but requires more changes to the code, and possibly semantics
  (although given no normalization was happening before, perhaps not much change).
* `tauto` sets flags to only recognize `Inductive Foo A B := foo : A -> B -> Foo A B` as conjunction.
  (the `strict` flags). The recorded reason for those flags is to only recognize types isomorphic to
  the standard connectives. In a future where `prod` becomes non-dependent `sigT`, but `and`
  remains unchanged, both patterns would be standard. What is the compatibility story for
  `tauto` vs `dtauto`?

I'll keep working on these changes, since they seem like they would be useful for `dtauto` at least.
